daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala

```diff
@@ -522,4 +558,84 @@ object Engine {
     StableConfig.copy(allowedLanguageVersions = LanguageVersion.DevVersions)
   )
 
+  private[engine] case class DisclosureTable(
+      allDisclosures: ImmArray[DisclosedContract],
+      contractIdByKey: Map[crypto.Hash, ContractId],
+      contractById: Map[ContractId, VersionedContractInstance],
+  )
+
+  private[engine] def buildDiscTable(
+      machine: Machine,
+      disclosures: ImmArray[speedy.DisclosedContract],
+  ): Result[DisclosureTable] = {
+    val acc = disclosures.foldLeft(
+      (
+        Map.empty[ContractId, VersionedContractInstance],
+        Map.empty[crypto.Hash, ContractId],
+        List.empty[DisclosedContract],
+        Option.empty[Error.Preprocessing.BadDisclosedContract],
+      )
+    ) { case ((m1, m2, ds, errorM), d) =>
+      errorM match {
+        case None =>
+          val arg = machine.normValue(d.templateId, d.argument)
+          val coid = machine.normValue(d.templateId, d.contractId)
+          val version = machine.tmplId2TxVersion(d.templateId)
+
+          // check for well typed contract argument
+          (coid, arg) match {
+            case (ValueContractId(coid), r @ ValueRecord(Some(d.templateId), _)) =>
+              // check for duplicate contract ids
+              m1.get(coid) match {
+                case Some(_) =>
+                  (
+                    m1,
+                    m2,
+                    ds,
+                    Some(Error.Preprocessing.BadDisclosedContract("Duplicate disclosed contract.")),
+                  )
+                case None =>
+                  val contractInst = VersionedContractInstance(
+                    template = d.templateId,
+                    arg = Versioned(version, r),
+                    agreementText = "",
+                  )
+                  val m1_prime = m1 + (coid -> contractInst)
+                  val d_prime = DisclosedContract(
+                    templateId = d.templateId,
+                    argument = arg,
+                    contractId = coid,
+                    metadata = d.metadata,
+                  )
+                  val ds_prime = d_prime :: ds
+                  d.metadata.keyHash match {
+                    case Some(hash) => (m1_prime, m2 + (hash -> coid), ds_prime, errorM)
+                    case None => (m1_prime, m2, ds_prime, errorM)
+                  }
+              }
+            case _ => (m1, m2, ds, errorM)
+          }
+        // abort on error.
+        case Some(_) =>
```

Comment:
why are you throwing away the error here and create a new one?

Reply:
I would have to use a monadic fold to process the error, and ImmArray doesn't provide that. As discussed below, I'll use a normal exception.

Reply:
replaced by an exception now.